### PR TITLE
Add a test for 5-9 yrs group

### PR DESCRIPTION
**Pull request by Wartynas**
Hey @HorridTom, I was able to run and fail the test with the reason for that being an incorrect location of 5-9 yrs age band. Please review it and let me know what should be changed.

### CODE REVIEW

**Comment by ImogenConnorHelleur** [comment]
Hi @Wartynas, test looks good! A few points to consider though:

- Probably want to remove the `print(class(test_data))`
- Also want to remove the reference to Chelsea & Westminster
- I don't think we need the `if` statement because we'll always be running the test on that same data set so there shouldn't ever be the case where the "5-9 yrs" category isn't there - and if it isn't there, something has probably gone wrong so we'd want the  test to fail.
- I think the correct answer should actually be 3 and not 2 because the 5-9 yrs" category comes after "0 yrs" and "1-4 yrs" 

Good first test though :)

**Comment by Wartynas** [comment]
Thanks @ImogenConnorHelleur! Completely agree with you. Will fix the test.

**Comment by Wartynas** [comment]
Hi @HorridTom and @ImogenConnorHelleur. Yesterday we discussed possible solutions of age band ordering problem and Imogen suggested setting the `age_band_start` variable as a `factor` with predefined `levels` of age groups (see [here](https://github.com/HorridTom/hospitalflow/blob/740ef0b903cccd0e41d04db0d4537e23c7806382/R/ae_attendances_admissions_age_sex.R#L32)). It worked like a charm - now a function call yield correct ordering and all 57 tests pass. I think the branches are ready to be merged, but please have a look and let me know if any changes are needed.

fixed

**Comment by HorridTom** [comment]
Hi @Wartynas @ImogenConnorHelleur thanks for this. I agree with the edits to the test. However, I don't think we should order the age bands in the code; I think this should be done on import through a config file. The reason is that in principle there is no reason why the age bands in a given dataset need to be exactly these - i.e. what if the age bands went 0-9, 10-19, 20-29 etc. Then this code will throw an error. Whereas a config file can be set up for each dataset, so that the order is set on import, the age bands display in the correct order, and the code can cope with whatever age bands are supplied. What do you think?

**Comment by Wartynas** [comment]
Hi @HorridTom. Thanks for your suggestion - I think this approach would definitely make the function more scalable and robust to changes. I see how this could be done, but there is one part that confuses me - how does the function 'know', which config file was used to create the dataset that is provided to the `data` argument of the function? Currently, the function expects a spell table to be provided as `data`, which is created from `ed` and `inpatient` tables, both of which had already been standartised (as mentioned in the issue [discussion](https://github.com/HorridTom/hospitalflow/issues/103#issuecomment-1038942498) by @ImogenConnorHelleur). To make use of the config that was used to standardise the original tables, I am thinking of including another parameter to the function, say, `path_to_age_cfg` or something similar that would allow the access of standard age bands. I must say I am not sure if this is what you had in mind and if it is an efficient approach at all, so I would like to hear your feedback before taking action.

**Comment by Wartynas** [comment]
Hi @HorridTom, I used `gtools::mixedsort` and `base::unique` functions to get factor levels from the values of the `age_band_start` column. What do you think of this approach?

File: R/ae_attendances_admissions_age_sex.R [hunk 1]
```diff
@@ -29,6 +29,7 @@ ae_attendances_admissions_age_sex <- function(start_date, end_date, data, plot_c
     dplyr::filter(spell_start <= end_date & spell_end >= start_date) %>%
     dplyr::select(spell_number, spell_start, spell_end, gender, age_band_start, spell_class_col)
 
+  dt$age_band_start <- factor(dt$age_band_start, levels = gtools::mixedsort(unique(dt$age_band_start)))
```
Comment (HorridTom) [inline R/ae_attendances_admissions_age_sex.R:32]: This looks great - if possible, could we use `dplyr::mutate` here? So
`dt <- dt %>% dplyr::mutate(age_band_start = factor(age_band_start, levels = ...as your code...))`

**Review by HorridTom — CHANGES_REQUESTED** [review]
This looks great, just a minor style comment  on using tidyverse functions for consistency

**Comment by HorridTom** [comment]
@Wartynas I like it - good idea! So this way, the data can have whatever age bands are provided by the hospital in question, but the age bands are shown in the right order on the graph - nice.

Some thoughts:

1. See my comment on the code line itself re. using mutate here
2. If gtools is a new package that we haven't used before in hospitalflow, we'll need to ensure a) that we update renv and b) that the description file is update (using use_package). We can do this together perhaps as it may be quite new to you?
3. Ultimately, it would be nice if this sorting of the age bands was dealt with on import, so that any spell table created by the package _must_ have age bands in the correct order - this then ensures that any other charts that display age bands also show them in the right order, without us having to fix the order in every charting function... but I think that is an issue for another day - perhaps you could add an issue for that so we don't forget?

File: R/ae_attendances_admissions_age_sex.R (hunk 1 above)
Comment (Wartynas) [reply R/ae_attendances_admissions_age_sex.R]: Yes, you should see the code rewritten with `dplyr` [here](https://github.com/HorridTom/hospitalflow/pull/105/commits/a4b136ccba8a5a77af5eb3bb18fd05863be6d1bf). I tried to put `dplyr::mutate(...` in line 31 and `%>%` at the end of line 30 (it makes the code more concise and avoids `dt <- dt %>%` line), but I got a strange error about closure being not subsettable, so left it like this.

**Comment by Wartynas** [comment]
Hey @HorridTom. Yes, that was the whole point.
1. Done.
2. `gtools` was not previously used in `hospitalflow`. Yes, it would be great if we could do those steps together for the first time.
3. I see what you mean. I guess this could be achieved by editing one of the functions in `spell_linkage.R`, a script that creates spell tables from ed and inpatient data? Will create an issue for that.

Also, could you please clarify what you mean by "on import" in the previous comment? I kind of intuitively grasp what you mean by that, but would not be able to say exactly.

**Comment by HorridTom** [comment]
Hi @Wartynas, this is now looking good - just one thing - when I run the tests, I'm now getting a failure of the other test in that same test file:
`Failure (test_attend_adm_age_sex.R:64:3): Admission and Attendances by age-Sex, for improvised data
result_data not equal to correct_answers_1.
Component "age_band_start": Attributes: < Component "levels": Lengths (2, 21) differ (string compare on first 2) >
Component "age_band_start": Attributes: < Component "levels": 2 string mismatches >`
You can see why if you run the lines of the test in the console, then run:
`result_data$age_band_start` and
`correct_answers_1$age_band_start`; the levels of the factor variables are different.

This is because the new functionality you have implemented means that the data returned from the function only has the levels actually used in the input data (in this case only two `Levels: 1-4 yrs 95+`), whereas the correct answer sets up the factor with all of the standard levels.

Thinking about this, I realised that there is a use case where only using the levels present in the data is problematic. Imagine some example data where the numbers are quite low, and there is one age band in the middle with 0 attendances and admissions; that age band is going to be missing from the graph, whereas we want it to be there, but with no bar. I think the solution, therefore, is to apply `gtools::mixedsort` to the _levels_ of the age_band_start factor of the input data (`dt`) rather than to the values of that factor. I'll add a comment to that line to be a bit clearer. But that should then preserve the levels from the input data in the output graph/table. The old test should then pass again...

File: R/ae_attendances_admissions_age_sex.R [hunk 2]
```diff
@@ -29,6 +29,8 @@ ae_attendances_admissions_age_sex <- function(start_date, end_date, data, plot_c
     dplyr::filter(spell_start <= end_date & spell_end >= start_date) %>%
     dplyr::select(spell_number, spell_start, spell_end, gender, age_band_start, spell_class_col)
 
+  dt <- dt %>%
+    dplyr::mutate(age_band_start = factor(age_band_start, levels = gtools::mixedsort(unique(dt$age_band_start)))) # necessary for correct age band ordering on x axis
```
Comment (HorridTom) [inline R/ae_attendances_admissions_age_sex.R:33]: @Wartynas I think this should now be ... `levels = tools::mixedsort(unique(levels(dt$age_band_start))` - see my comment on the main thread of the PR

Comment (HorridTom) [reply R/ae_attendances_admissions_age_sex.R:33]: Ah no hold on - I've just realised at this point age_band_start is not a factor... bear with me...

**Comment by HorridTom** [comment]
OK @Wartynas I can't see what I'm missing on this - I'm getting a failing test as above. Are you getting that too? But now when I run result_data$age_band_start
result$age_band_start and
correct_answers_1$age_band_start; they are the same, and I cannot find any differences at all, and yet the test is still failing... I have to stop now, but will revisit this tomorrow. If you could let me know if you get the same failing test that would be useful... Thanks

**Comment by Wartynas** [comment]
I did `levels = gtools::mixedsort(unique(levels(dt$age_band_start))` in the `ae_attendances_admissions_age_sex` function (line 33) and this solved the problem. However, now the test I wrote for 5-9 years band does not pass. Here is the error:
```
Failure (test_attend_adm_age_sex.R:80:3): 5-9 yrs age group is in the correct place
which("5-9 yrs" == labels) not equal to 3.
Lengths differ: 0 is not 1
```

Maybe we should get rid of this test, since `gtools:mixedsort` ensures correct order?

**Comment by HorridTom** [comment]
Yeah I think you had right before, sorry my bad - you can ignore my suggestion to add the "levels" call in - that doesn't work because at that point in the code the variable is not a factor, it's character.

But we should not get rid of the first test, it is testing that the actual height of the bars are correctly calculated. It's very puzzling because when I run the code in the console, the results and correct_answers_1 objects seem identical... so I don't know why that test is failing!

I'd like to investigate further... but if you can figure it out, let me know!

**Comment by HorridTom** [comment]
Ok @Wartynas I think I have worked it out... In the test data that is "manually" created in the first test ("Admission and Attendances by age-Sex, for improvised data") the age_band_start column is a factor. In the test data you load in in your new test, it is character. I think I have a fix that handles both cases - let me know if ok for me to commit to this branch and I'll commit my suggestion for you to review. (Basically ok for me to commit if you have a "clean" state on the branch i.e. no uncommitted changes).

**Comment by Wartynas** [comment]
Hi @HorridTom, the branch has no uncommitted changes. Please commit your suggestion and I will check it as soon as you do it. Thank you very much for your help!

**Comment by HorridTom** [comment]
Hi @Wartynas I've pushed those changes. This is the commit that makes the difference: [af01dcf](https://github.com/HorridTom/hospitalflow/pull/105/commits/af01dcf42631609038d72e56615aa95e0174a565)

Let know if you agree this fixes things, until we address #107 at least.

**Comment by Wartynas** [comment]
Hi @HorridTom, I pulled the changes and ran the tests - 57/57 pass. I also ran the function with different datasets - all cases seem to produce expected results, with "empty" age bands included in the plot. 

Although I have a question about [this](https://github.com/HorridTom/hospitalflow/blob/13a7b667335cdca0dfff054935b7f480d7cb21bc/R/ae_attendances_admissions_age_sex.R#L33) line in the if-else statement. Is it necessary to wrap `levels(dt$age_band_start)` with `unique` function like this `unique(levels(dt$age_band_start))`? I am asking this because the output of `levels` will always (?) be a vector containing unique values.

**Comment by HorridTom** [comment]
> Hi @HorridTom, I pulled the changes and ran the tests - 57/57 pass. I also ran the function with different datasets - all cases seem to produce expected results, with "empty" age bands included in the plot.
Great - nice one.

> Although I have a question about [this](https://github.com/HorridTom/hospitalflow/blob/13a7b667335cdca0dfff054935b7f480d7cb21bc/R/ae_attendances_admissions_age_sex.R#L33) line in the if-else statement. Is it necessary to wrap `levels(dt$age_band_start)` with `unique` function like this `unique(levels(dt$age_band_start))`? I am asking this because the output of `levels` will always (?) be a vector containing unique values.
You are absolutely right yes - you ok to edit that?

One final thing; I wonder if we might use a name other than `plot` for the plot object you create in the test - just in general it is good to avoid making objects with the same name as base/common R functions...

If you make those changes, we can merge in tomorrow when we meet?

**Pull request merged**